olmo/train.py

```diff
@@ -807,14 +856,35 @@ def train_batch(self, batch: Dict[str, Any]) -> Tuple[torch.Tensor, Optional[tor
                         assert z_batch_loss is not None
                         z_batch_loss += z_loss.detach()
 
+                if self.model.config.block_type == BlockType.moe:
+                    if self.model.config.moe_zloss_weight is not None:
+                        lb_loss, moe_z_loss = batched_load_balancing_loss(self.moe_args)
+                        lb_loss = lb_loss / len(micro_batches)
+                        moe_z_loss = moe_z_loss / len(micro_batches)
+                    else:
+                        lb_loss = batched_load_balancing_loss(self.moe_args) / len(micro_batches)
+                    if self.model.config.moe_log_expert_assignment:
+                        if self.model.config.moe_zloss_weight:
+                            tokens_per_expert, _, _ = zip(*get_load_balancing_loss())
+                        else:
+                            tokens_per_expert, _ = zip(*get_load_balancing_loss())
+                        expert_assignments += torch.stack(tokens_per_expert, dim=0).cpu()
```

Comment:
Not a big deal, but are you sure that this is faster on CPU? Back in the day I always thought this too, keep small stuff on the CPU, but in practice doing it all on GPU was always faster.

Reply:
If `tokens_per_expert` were on GPU this will trigger a host-device sync. In that case it's almost certainly better to keep on GPU.

Reply:
> were on GPU this will trigger a host-device sync

Don't we want to avoid these syncs?

Reply:
Moved it to GPU now